plugin/forward/connect.go

```diff
@@ -44,29 +44,30 @@ func (t *Transport) updateDialTimeout(newDialTime time.Duration) {
 }
 
 // Dial dials the address configured in transport, potentially reusing a connection or creating a new one.
-func (t *Transport) Dial(proto string) (*dns.Conn, bool, error) {
+func (t *Transport) Dial(proto string) (*persistConn, bool, error) {
 	// If tls has been configured; use it.
 	if t.tlsConfig != nil {
 		proto = "tcp-tls"
 	}
 
 	t.dial <- proto
-	c := <-t.ret
+	pc := <-t.ret
 
-	if c != nil {
-		return c, true, nil
+	if pc != nil {
+		pc.used = time.Now() // update used time
```

Comment:
If we update `used` at dial time this doesn't guarantee that connections will be yielded in the same order. Ideally, `connManager()` should (and actually does) update time at yield time. As for me, I would keep the existing logic.
But I guess you want to move as much code as possible from connManager() loop. That's why I suggested to update `used` in `func (t *Transport) Yield(pc *persistConn)`. This could make cached connections slightly unordered, though.